assets/js/blocks/cart-checkout/cart/full-cart/index.tsx

```diff
@@ -114,6 +114,18 @@ const Cart = ( { attributes }: CartProps ) => {
 		cart,
 	};
 
+	const discountsSlotFillProps = {
+		extensions,
+		cart,
+		contexts: {
+			validation: useValidationContext(),
+			storeCart: useStoreCart(),
+		},
+		components: {
+			ValidationInputError,
+		},
+	};
```

Comment:
Instead of doing this which is a side-effect of exporting `ValidatedTextInput` from the checkout package (which wouldn't have access to the same validation context instance). Could you avoid removing the encapsulation of the validation logic in `ValidatedTextInput` and have the component passed through on the slot-fill here? So you'd end up with something like:

```suggestion
	const discountsSlotFillProps = {
		extensions,
		cart,
		contexts: {
			storeCart: useStoreCart(),
		},
		components: {
			ValidatedTextInput
		},
	};
```

The concern I have with the current implementation is it exposes more of the implementation details to extensions and makes it harder for us to make changes in the future.

Reply:
I also have a high level of concern around exposing the entire cart implementation details here. I realize points and rewards redemption requires some ability to update the cart but this again exposes implementation details that makes it harder for us to modify how this works in the future - especially when in the [current implementation](426-gh-woocommerce/woocommerce-points-and-rewards/files#diff-887f8916eac678e0237153f2b3c99f800a46465825202cb0e4f1d30d713cfb28R35-R39) there's some pretty high dependency on [making sure the right shape for the cart data](426-gh-woocommerce/woocommerce-points-and-rewards/files#diff-184b109d487795364c256a34343c3fa20257a671debff646fa7e4bc31ebbe638R202) is fed to the `receiveCart` dispatcher.

This is an interface that would be fairly hard for extensions to understand and apply correctly without deep knowledge of internals.

I don't have any immediate alternatives to suggest but I'd be wary about shipping with the current implementation.

Reply:
Thanks Darren, I'll work on the `ValidatedTextInput` change, it should be fine, and actually, right now I no longer need the `useStoreCart` context, following the addition of https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4248/files#diff-e9030d5520892c9926d801219c89bbc21b4863a510ce18409f74b0a148f1270eR1-R24

BUT, the caveat is, after updating the cart, the extension needs to request fresh data to be loaded, which has a pretty terrible UX especially on slower devices or connections: https://www.loom.com/share/8a9871aeb0b5494a84cfc54f84c00dfd

Reply:
> Are you suggesting to move ValidatedTextInput back out of the checkout package and instead pass it through the slot?
Yes this.

Reply:
> BUT, the caveat is, after updating the cart, the extension needs to request fresh data to be loaded, which has a pretty poor UX especially on slower devices or connections: https://www.loom.com/share/8a9871aeb0b5494a84cfc54f84c00dfd

Ya that's not the best UX. I wonder if instead of doing this, we could provide some API that signals the checkout should do some sort of "update_cart" post to the server on the StoreAPI that extensions could react to server side and make any necessary updates? So it'd work similarly to when quantity is updated etc except extensions are simply triggering the post and response? So the rough flow would be something like:

1. Extension updates a value attached to it's state in the cart store.
2. Extension uses some way of signaling it's state needs updated in on the server.
3. This triggers a checkout "update" POST.
4. Extension is hooked into the StoreAPI "update" POST for capturing the data it needs to process server side and preparing the response.
5. StoreAPI returns the updated Cart response (and the extension could react to value changes in it's cart store state)?

Reply:
Thanks for your feedback @nerrad, as we have discussed in our 1:1 and in [this PR](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4298), this is the approach I think we'll continue with. 

I have reverted the changes in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4238 and instead opted to pass this through the slot, due to us not wanting to pass validation contexts through the slot.

This PR is now ready for a full review 🎉